File: packages/cli/hieudoanm/clap.rs/src/libs/history.rs
use chrono::Local;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::fs;
use std::path::PathBuf;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Entry {
    pub timestamp: String,
    pub source: String,
    pub command: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub cwd: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub duration_ms: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

#[derive(Debug, Serialize)]
pub struct CommandCount {
    pub name: String,
    pub count: usize,
}

#[derive(Debug, Serialize)]
pub struct Stats {
    pub total_cli: usize,
    pub total_mcp: usize,
    pub top_commands: Vec<CommandCount>,
    pub top_errors: Vec<CommandCount>,
}

fn storage_path() -> anyhow::Result<PathBuf> {
    let home = dirs::home_dir().ok_or_else(|| anyhow::anyhow!("cannot find home dir"))?;
    let dir = home.join(".hieudoanm");
    fs::create_dir_all(&dir)?;
    Ok(dir.join("history.jsonl"))
}

pub fn append(entry: &Entry) -> anyhow::Result<()> {
    let path = storage_path()?;
    let line = serde_json::to_string(entry)?;
    let data = format!("{}\n", line);
    use std::io::Write;
    let mut f = fs::OpenOptions::new()
        .append(true)
        .create(true)
        .open(&path)?;
    f.write_all(data.as_bytes())?;
    Ok(())
}

fn read_all() -> anyhow::Result<Vec<Entry>> {
    let path = storage_path()?;
    if !path.exists() {
        return Ok(Vec::new());
    }
    let content = fs::read_to_string(&path)?;
    let mut entries = Vec::new();
    for line in content.lines() {
        let line = line.trim();
        if line.is_empty() {
            continue;
        }
        if let Ok(e) = serde_json::from_str::<Entry>(line) {
            entries.push(e);
        }
    }
    Ok(entries)
}

pub fn list(count: usize) -> anyhow::Result<Vec<Entry>> {
    let entries = read_all()?;
    let n = if count == 0 || count > entries.len() {
        entries.len()
    } else {
        count
    };
    Ok(entries[entries.len() - n..].to_vec())
}

pub fn search(query: &str, limit: usize) -> anyhow::Result<Vec<Entry>> {
    let entries = read_all()?;
    let q = query.to_lowercase();
    let mut results = Vec::new();
    for e in entries.into_iter().rev() {
        if e.command.to_lowercase().contains(&q) {
            results.push(e);
            if limit > 0 && results.len() >= limit {
                break;
            }
        }
    }
    Ok(results)
}

pub fn clear() -> anyhow::Result<()> {
    let path = storage_path()?;
    if path.exists() {
        fs::remove_file(&path)?;
    }
    Ok(())
}

pub fn compute_stats() -> anyhow::Result<Stats> {
    let entries = read_all()?;
    let mut stats = Stats {
        total_cli: 0,
        total_mcp: 0,
        top_commands: Vec::new(),
        top_errors: Vec::new(),
    };
    let mut cmd_count: HashMap<String, usize> = HashMap::new();
    let mut err_count: HashMap<String, usize> = HashMap::new();
    for e in &entries {
        if e.source == "cli" {
            stats.total_cli += 1;
        } else {
            stats.total_mcp += 1;
        }
        *cmd_count.entry(e.command.clone()).or_insert(0) += 1;
        if e.error.is_some() {
            *err_count.entry(e.command.clone()).or_insert(0) += 1;
        }
    }
    stats.top_commands = top_n(&cmd_count, 10);
    stats.top_errors = top_n(&err_count, 10);
    Ok(stats)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_top_n_empty() {
        let counts = HashMap::new();
        let result = top_n(&counts, 5);
        assert!(result.is_empty());
    }

    #[test]
    fn test_top_n_basic() {
        let mut counts = HashMap::new();
        counts.insert("a".to_string(), 10);
        counts.insert("b".to_string(), 5);
        counts.insert("c".to_string(), 20);
        let result = top_n(&counts, 5);
        assert_eq!(result.len(), 3);
        assert_eq!(result[0].name, "c");
        assert_eq!(result[0].count, 20);
        assert_eq!(result[2].name, "b");
        assert_eq!(result[2].count, 5);
    }

    #[test]
    fn test_top_n_limits_results() {
        let mut counts = HashMap::new();
        for i in 0..10 {
            counts.insert(format!("cmd{i}"), i);
        }
        let result = top_n(&counts, 3);
        assert_eq!(result.len(), 3);
    }

    #[test]
    fn test_entry_display_no_error() {
        let e = Entry {
            timestamp: "2025-01-01".to_string(),
            source: "cli".to_string(),
            command: "test".to_string(),
            cwd: None,
            duration_ms: None,
            error: None,
        };
        let output = format!("{e}");
        assert_eq!(output, "2025-01-01  test");
    }

    #[test]
    fn test_entry_display_with_error() {
        let e = Entry {
            timestamp: "2025-01-01".to_string(),
            source: "cli".to_string(),
            command: "test".to_string(),
            cwd: None,
            duration_ms: None,
            error: Some("failed".to_string()),
        };
        let output = format!("{e}");
        assert_eq!(output, "2025-01-01  test  [failed]");
    }

    #[test]
    fn test_entry_basic() {
        let e = entry("cli", "help");
        assert_eq!(e.source, "cli");
        assert_eq!(e.command, "help");
        assert!(e.timestamp.len() > 10);
    }

    #[test]
    fn test_stats_totals() {
        use std::collections::HashMap;
        let counts: HashMap<String, usize> = HashMap::new();
        let result = top_n(&counts, 5);
        assert!(result.is_empty());
    }
}

fn top_n(counts: &HashMap<String, usize>, n: usize) -> Vec<CommandCount> {
    let mut v: Vec<CommandCount> = counts
        .iter()
        .map(|(k, v)| CommandCount {
            name: k.clone(),
            count: *v,
        })
        .collect();
    v.sort_by_key(|b| std::cmp::Reverse(b.count));
    v.truncate(n);
    v
}

impl std::fmt::Display for Entry {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}  {}", self.timestamp, self.command)?;
        if let Some(ref err) = self.error {
            write!(f, "  [{}]", err)?;
        }
        Ok(())
    }
}

pub fn entry(source: &str, command: &str) -> Entry {
    Entry {
        timestamp: Local::now().format("%Y-%m-%dT%H:%M:%S%:z").to_string(),
        source: source.to_string(),
        command: command.to_string(),
        cwd: std::env::current_dir()
            .ok()
            .map(|p| p.to_string_lossy().to_string()),
        duration_ms: None,
        error: None,
    }
}

File: packages/cli/hieudoanm/cobra.go/src/libs/history/history.go
package history

import (
	"encoding/json"
	"os"
	"path/filepath"
	"sort"
	"strings"
)

type Entry struct {
	Timestamp  string `json:"timestamp"`
	Source     string `json:"source"`
	Command    string `json:"command"`
	CWD        string `json:"cwd,omitempty"`
	DurationMs int64  `json:"duration_ms,omitempty"`
	Error      string `json:"error,omitempty"`
}

type CommandCount struct {
	Name  string `json:"name"`
	Count int    `json:"count"`
}

type Stats struct {
	TotalCLI    int            `json:"total_cli"`
	TotalMCP    int            `json:"total_mcp"`
	TopCommands []CommandCount `json:"top_commands"`
	TopErrors   []CommandCount `json:"top_errors"`
}

func storagePath() (string, error) {
	home, err := os.UserHomeDir()
	if err != nil {
		return "", err
	}
	dir := filepath.Join(home, ".hieudoanm")
	if err := os.MkdirAll(dir, 0755); err != nil {
		return "", err
	}
	return filepath.Join(dir, "history.jsonl"), nil
}

func Append(entry Entry) error {
	path, err := storagePath()
	if err != nil {
		return err
	}
	data, err := json.Marshal(entry)
	if err != nil {
		return err
	}
	f, err := os.OpenFile(path, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
	if err != nil {
		return err
	}
	defer f.Close()
	_, err = f.Write(append(data, '\n'))
	return err
}

func readAll() ([]Entry, error) {
	path, err := storagePath()
	if err != nil {
		return nil, err
	}
	data, err := os.ReadFile(path)
	if err != nil {
		if os.IsNotExist(err) {
			return nil, nil
		}
		return nil, err
	}
	var entries []Entry
	for _, line := range strings.Split(string(data), "\n") {
		line = strings.TrimSpace(line)
		if line == "" {
			continue
		}
		var e Entry
		if err := json.Unmarshal([]byte(line), &e); err != nil {
			continue
		}
		entries = append(entries, e)
	}
	return entries, nil
}

func List(count int) ([]Entry, error) {
	entries, err := readAll()
	if err != nil {
		return nil, err
	}
	if count <= 0 || count > len(entries) {
		count = len(entries)
	}
	return entries[len(entries)-count:], nil
}

func Search(query string, limit int) ([]Entry, error) {
	entries, err := readAll()
	if err != nil {
		return nil, err
	}
	q := strings.ToLower(query)
	var results []Entry
	for i := len(entries) - 1; i >= 0; i-- {
		e := entries[i]
		if strings.Contains(strings.ToLower(e.Command), q) {
			results = append(results, e)
			if limit > 0 && len(results) >= limit {
				break
			}
		}
	}
	return results, nil
}

func Clear() error {
	path, err := storagePath()
	if err != nil {
		return err
	}
	if err := os.Remove(path); err != nil && !os.IsNotExist(err) {
		return err
	}
	return nil
}

func ComputeStats() (*Stats, error) {
	entries, err := readAll()
	if err != nil {
		return nil, err
	}
	stats := &Stats{}
	cmdCount := make(map[string]int)
	errCount := make(map[string]int)
	for _, e := range entries {
		if e.Source == "cli" {
			stats.TotalCLI++
		} else {
			stats.TotalMCP++
		}
		cmdCount[e.Command]++
		if e.Error != "" {
			errCount[e.Command]++
		}
	}
	stats.TopCommands = topN(cmdCount, 10)
	stats.TopErrors = topN(errCount, 10)
	return stats, nil
}

func topN(counts map[string]int, n int) []CommandCount {
	var result []CommandCount
	for name, count := range counts {
		result = append(result, CommandCount{Name: name, Count: count})
	}
	sort.Slice(result, func(i, j int) bool {
		return result[i].Count > result[j].Count
	})
	if len(result) > n {
		result = result[:n]
	}
	return result
}

func (e Entry) String() string {
	var b strings.Builder
	b.WriteString(e.Timestamp)
	b.WriteString("  ")
	b.WriteString(e.Command)
	if e.Error != "" {
		b.WriteString("  [")
		b.WriteString(e.Error)
		b.WriteString("]")
	}
	return b.String()
}
